### PR TITLE
expression: Report error when empty pattern in regexp related function

### LINKED ISSUE #53221: Regexp should report error when pattern parameter is empty string
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
mysql> create table t0 (a varchar(10));
mysql> insert into t0 values('');
mysql> select 'hello' regexp a from t0;

### 2. What did you expect to see? (Required)
Error "Empty pattern is invalid" should be reported
### 3. What did you see instead (Required)
```
mysql> select 'hello' regexp a from t0;
+------------------+
| 'hello' regexp a |
+------------------+
|                1 |
+------------------+
1 row in set (0.01 sec)
```
### 4. What is your TiDB version? (Required)
---------------------------------------------------------------------------------------+
| Release Version: v8.2.0-alpha-6-g8f138d5113
Edition: Community
Git Commit Hash: 8f138d511325aaafd0ba7e94d1f6fb1dc4ac7b36
Git Branch: HEAD
UTC Build Time: 2024-04-22 08:00:01
GoVersion: go1.21.6
Race Enabled: false
Check Table Before Drop: false
Store: tikv |

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: close #53221 

Problem Summary:
Tidb report error when these functions are executed in row by row mode. But failed to check it in vector mode.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/expression/builtin_regexp.go [hunk 1]
```diff
@@ -583,8 +587,11 @@ func (re *builtinRegexpSubstrFuncSig) vecEvalString(ctx EvalContext, input *chun
 
 		if !memorized {
 			// Get pattern and match type and then generate regexp
-			pattern := params[1].getStringVal(i)
 			matchType := params[4].getStringVal(i)
+			pattern := params[1].getStringVal(i)
```
Comment (xzhangxian1008) [inline pkg/expression/builtin_regexp.go:591]: Maybe we can check empty pattern in `buildRegexp`?

Comment (yibin87) [reply pkg/expression/builtin_regexp.go:591]: Nice suggestion, I'll change it

**Comment by yibin87** [comment]
hold to wait for copr-test pr https://github.com/tikv/copr-test/pull/183 merged

File: pkg/expression/builtin_regexp.go [hunk 2]
```diff
@@ -314,7 +318,8 @@ func (re *builtinRegexpLikeFuncSig) vecEvalInt(ctx EvalContext, input *chunk.Chu
 
 		if !memorized {
 			matchType := params[2].getStringVal(i)
-			reg, err = re.buildRegexp(params[1].getStringVal(i), matchType)
+			pattern := params[1].getStringVal(i)
+			reg, err = re.buildRegexp(pattern, matchType)
```
Comment (windtalker) [inline pkg/expression/builtin_regexp.go:322]: Don't need to change this anymore?

Comment (yibin87) [reply pkg/expression/builtin_regexp.go:322]: Use local variable here looks all right also, since matchType is used here too.

**Review by xzhangxian1008 — APPROVED** [review]
LGTM

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**